File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    private boolean dfsCheck(int node,ArrayList<ArrayList<Integer>> adj,int vis[],int pathVis[],Stack<Integer> st){
        vis[node]=1;
        pathVis[node]=1;

        for(int it: adj.get(node)){
            if(vis[it]==0){
                if(dfsCheck(it,adj,vis,pathVis,st)==true) return true;
            }else if(pathVis[it]==1){
                return true;
            }
        }
        pathVis[node]=0;
        st.push(node);
        return false;

    }
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int arr[] = new int[numCourses];
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for(int i=0;i<numCourses;i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0;i<prerequisites.length;i++){
            adj.get(prerequisites[i][1]).add(prerequisites[i][0]);
        }

        int vis[] = new int[numCourses];
        int pathVis[] = new int[numCourses];
        Stack<Integer> st = new Stack<>();
        for(int i=0;i<numCourses;i++){
            if(vis[i]==0){
                if(dfsCheck(i,adj,vis,pathVis,st)==true){
                    return new int[]{};
                }
            }
        }
        int idx =0;
        while(!st.isEmpty()){
            arr[idx++] = st.pop();
        }
        return arr;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
    private:
    bool dfsCheck(int node, vector<vector<int>>& adj,vector<int>& vis,vector<int>& pathVis,stack<int>& st){
        vis[node]=1;
        pathVis[node]=1;
        for(auto it: adj[node]){
            if(!vis[it]){
                if(dfsCheck(it,adj,vis,pathVis,st)==true) return true;
            }else if(pathVis[it]==1) return true;
        }

        pathVis[node]=0;
        st.push(node);
        return false;
    }
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> vis(numCourses,0);
        vector<int> pathVis(numCourses,0);
        vector<vector<int>> adj(numCourses);
        vector<int> ans;
        for(int i=0;i<prerequisites.size();i++){
            adj[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        std::stack<int> st;
        for(int i=0;i<numCourses;i++){
            if(!vis[i]){
                if(dfsCheck(i,adj,vis,pathVis,st)==true) return {};
            }
        }

        while(!st.empty()){
            ans.push_back(st.top());
            st.pop();
        }
        return ans;
    }
};